### ISSUE TITLE
(Placeholder) Design a way to reliably create stable hashes for cards

### ISSUE THREAD

**Issue by maxammann** [digitalfabrik/entitlementcard#673]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by sarahsporck** [comment]
Hmm in the protobuf docu it is explicitly mentioned that the toJson is not considered as a stable implementation. Therefore, I'd consider using the first approach.

In that regard I have some questions: 
- The version field should be added to the protobuf, right? As in #699 you talk about adding a version to the hash.
- What do you mean by check that all fields are used? Does that mean every QrCode must contain every extension? Then why are the extensions optional. I feel like I'm missing sth. here.

**Comment by michael-markl** [comment]
> Hmm in the protobuf docu it is explicitly mentioned that the toJson is not considered as a stable implementation.

I am not too sure about that. They do state that they follow the [proto3 json spec](https://developers.google.com/protocol-buffers/docs/proto3#json) which clearly states which fields are encoded in which way. However, I agree that the output might still be too unstable (@maxammann what do you think?).
However, I would not discard the JSON idea fully, yet.

> The version field should be added to the protobuf, right?

If we should ever need a different hashing method (which we should aim to avoid imo), only then we'd need a version.
We can ignore that task until we actually need it, though.

> What do you mean by check that all fields are used? 

**Disclaimer: The issue description is probably a bit outdated; A solution still needs to be discussed.**
Previously, our approach was just to concatenate all non-message fields (using a predetermined order) encoded in binary; if fields were "missing", some defaults were used (zeros everywhere).
This is okay as long as no fields are added to the spec. Once that happens, we would have to write a new version of the algorithm with the added fields while keeping the old implementation for codes that were generated before that change.

To avoid this highly complicated scenario, we might want to make use of the optional fields and make the hashing algorithm "extensible" instead of maintaining multiple versions.
My suggestion for the algorithm specification:

---

**Procedure hashCardInfo(cardInfo, pepper)**
1. Implement a simple protobuf to json encoding (JSONEncode):
    1. **Messages** should be encoded as JSON-objects:
           If the message contains unknown fields, throw an error [In Flutter, the user should get a message like "Code konnte nicht gelesen werden. Ist die App auf dem neuesten Stand?"]
           For each of its fields, if ([field type](https://protobuf.dev/programming-guides/proto3/#specifying-field-rules) is `optional` and hasField()) or if field type is `singular` then add an entry with key=fieldTagNumber and value=JSONEncode(fieldValue)
           else throw *(for field types `repeated` and `map`)*
    3. **Strings** should be encoded as JSON-strings.
    4. **Integers** (int32,sint32,uint32,int64) should be encoded as a JSON-string (because JSON does not allow numbers in the int64 range)
    5. **Enums** should be encoded as a JSON-number
    6. If we encounter any other type (such as byte sequences, floats, etc.), throw
2. json = JSONEncode(cardInfo)
3. canonicalJson = CanonicalizeJson(json) // Returns a utf8 encoded byte array
4. return hash(canonicalJson, pepper)

---

Benefits: We could still make the following adjustments to the CardInfo proto without having to keep track of the specific version / adjust the hashing algorithm:
- Rename any fields
- Add a new (possibly top level) optional field
- Remove an optional field if it was never used

We can't do the following changes:
- Add a new (top level) singular field
- Make an optional field singular
- Make a singular field optional
- Remove an optional field if it was ever used

**Comment by sarahsporck** [comment]
@michael-markl can I shortly come back to this issue with you? I'm currently looking at #1147, and there we change the naming of 2 protobuf fields and one protobuf message.
I was wondering if this could be an issue for us, but as I see it, we should be completely free to change the names of fields, as they are not used at all in our protobufToJson implementation, right?

**Comment by michael-markl** [comment]
I replied in https://github.com/digitalfabrik/entitlementcard/pull/1147#issuecomment-1759448313.

TLDR: You can rename fields and messages of protobuf definitions without affecting the cardinfo hash nor the binary representation. You might still consider alternatives (such as introducing a new field / message) to avoid populating the same field with different data.
